src/model/element-reference.ts

```diff
@@ -0,0 +1,47 @@
+import {Resolvable, SourceRange, Feature} from '../model/model';
+import {Warning} from '../warning/warning';
+import * as dom5 from 'dom5';
+
+export interface Attribute {
+  name: string;
+  sourceRange: SourceRange|undefined;
```

Comment:
We should be able to type sourceRange and nameSourceRange without `undefined`. Just add `!` after the relevant expressions when you're creating the Attribute values. Only `valueSourceRange` really can be undefined, as far as I'm aware.